src/inference/tempering.jl

```diff
@@ -0,0 +1,178 @@
+###########################
+#   SAMPLING
+###########################
+
+"""
+Upon calling sample on a `TemperedSampler`, we must manually instantiate a `Sampler`
+to insert into the `TemperedSampler`, rather than wrapping the `TemperedSampler`
+itself in a `Sampler`
+"""
+function AbstractMCMC.sample(
+    rng::AbstractRNG,
+    model::AbstractModel,
+    sampler::MCMCTempering.TemperedSampler{<:InferenceAlgorithm},
+    N::Integer;
+    chain_type=MCMCChains.Chains,
+    resume_from=nothing,
+    progress=PROGRESS[],
+    kwargs...
+)
+    sampler = MCMCTempering.TemperedSampler(
+        Sampler(sampler.internal_sampler, model),
+        sampler.Δ,
+        sampler.Δ_init,
+        sampler.N_swap,
+        sampler.swap_strategy
+    )
+
+    if resume_from === nothing
+        return AbstractMCMC.mcmcsample(rng, model, sampler, N;
```

Comment:
I guess it would be more natural to call `sample` (which might fall back to `mcmcsample` but this should be defined by MCMCTempering):
```suggestion
        return AbstractMCMC.sample(rng, model, sampler, N;
```

Reply:
The problem I see here is that if you look at `Inference.jl`, the `Turing` specific AMCMC sample calls expect a `Sampler` subtyping `InferenceAlgorithm`, so we cannot call `sample` with a `TemperedSampler` and still pass through anything Turing specific here, hence why I just rewrote the final step of the chain so to speak but injecting the `Sampler` into the `internal_sampler` of the `TemperedSampler`.

Reply:
Sorry, I don't understand. If you call `mcmcsample` you circumvent any implementation of `sample` for `MCMCTempering.TemperedSampler` and just call the fallback definition in AbstractMCMC directly which you would hit anyway if `sample` is not implemented for these types. So it's still not clear to me what you gain from calling `mcmcsample` here apart from possibly missing a special implementation of `sample` for these types.